### ISSUE TITLE
pytest unable to import db fixture, yet lists it as available option

### ISSUE THREAD

**Issue by hancush** [pytest-dev/pytest-django#574]
I'm running Python 3.5.2, pytest 3.4.0, pytest-django 3.1.2, and Django 1.10.

I have a test directory structured like this:

```
tests
├── __init__.py
├── conftest.py (global fixtures)
├── submodule
│   ├── __init__.py
│   ├── conftest.py (module-specific fixtures)
│   ├── tests
├── submodule
│   ├── __init__.py
│   ├── conftest.py (module-specific fixtures)
│   ├── tests
└── test_config.py
```

I am able to include the `transactional_db` fixture for transactional database access in a fixture without issue. But separately, when I include the `db` fixture for non-transactional database access in a fixture, and run the tests with `pytest`, I get this error:

```
file /Users/Hannah/hackz/datamade/large-lots/tests/lots_client/test_eds.py, line 6
  @pytest.mark.django_db
  def test_eds_submission(application, client):
file /Users/Hannah/.virtualenvs/large-lots/lib/python3.5/site-packages/pytest_django/plugin.py, line 363
  @pytest.fixture(autouse=True)
  def _django_db_marker(request):
file /Users/Hannah/.virtualenvs/large-lots/lib/python3.5/site-packages/pytest_django/fixtures.py, line 141
  @pytest.fixture(scope='function')
  def db(request, django_db_setup, django_db_blocker):
file /Users/Hannah/hackz/datamade/large-lots/tests/lots_client/conftest.py, line 14
  @pytest.fixture
  @pytest.mark.django_db
  def django_db_setup(django_db_setup, db):
E       fixture 'db' not found
>       available fixtures: _dj_autoclear_mailbox, _django_clear_site_cache, _django_db_marker, _django_set_urlconf, _django_setup_unittest, _fail_for_invalid_template_variable, _live_server_helper, _template_string_if_invalid_marker, admin_client, admin_user, application, cache, capfd, capfdbinary, caplog, capsys, capsysbinary, client, db, django_db_blocker, django_db_keepdb, django_db_modify_db_settings, django_db_modify_db_settings_xdist_suffix, django_db_setup, django_db_use_migrations, django_test_environment, django_user_model, django_username_field, doctest_namespace, live_server, mailoutbox, monkeypatch, pytestconfig, record_xml_property, recwarn, rf, settings, tmpdir, tmpdir_factory, transactional_db
>       use 'pytest --fixtures [testpath]' for help on them.
```

Notice that `db` is among the available fixtures. It's also included when I run `pytest --fixtures [testpath]`. 

I've patched the problem by importing the `db` fixture directly from `pytest_django`, but I'm curious why this issue is occurring, and whether it's a bug or I've set something up incorrectly.

Thanks for your help!

**Comment by blueyed** [comment]
Maybe it's because you are using `@pytest.mark.django_db` also?
And you should not have to include `db` when overriding `django_db_setup`.

`db` is meant to be used in fixtures using the DB, but `django_db_setup` is internal, and meant for advanced use.

What are you trying to do there?
You should probably use it like this:

```python
@pytest.fixture
def django_db_setup(django_db_setup):
    …
```

**Comment by blueyed** [comment]
Would be nice to have something to reproduce this actually.
The best would be a failing test case for pytest-django itself - maybe we could improve the error in this case (assuming that you are trying to do something in a wrong way).

**Comment by hancush** [comment]
@blueyed thanks for getting back to me! taking a closer look, i think my problem was almost certainly a misunderstanding of the `django_db_setup` fixture. 

i have a series of factory fixtures that create a model object with override-able default data. obviously, these factory fixtures require db access and thus the `db` fixture. 

```python
@pytest.fixture
@pytest.mark.django_db
def factory_fixture(db)
    class Factory(object)
        def build(self, **kwargs):
            default_data = {'foo': 'bar'}
            default_data.update(kwargs)
            return SomeModel.objects.create(**default_data)
    return Factory()
``` 

i was trying to extend the `django_db_setup` fixture to create default data for a test module, using one of these factory fixtures.

```python
@pytest.fixture
def django_db_setup(django_db_setup, factory_fixture):
    factory_fixture.build(foo='custom foo')
```

i'm not 100% on the order of operation, but it seems like the `db` fixture is not available to the factory fixture, or any fixture, requested by the extended `django_db_setup`. in other words, i should not have mixed the lower level `django_db_*` fixtures with higher level fixtures like `db`. if i use the `django_db_blocker` fixture / `unblock` in the factory fixture, for example, things work as expected.

in any case, i think i've solved my own problem. it might be helpful to make a clearer distinction in the documentation between higher and lower level database fixtures, and maybe to raise an exception about mixed use being incompatible, but it's super possible this is an uncommon mistake on my part. 🙃

**Comment by blueyed** [comment]
Thanks for reporting back.
I'm glad you figured it out.
I've seen you creating https://github.com/datamade/testing-guidelines/issues/21 and just wanted to mention that it would be nice to have this in pytest-django's docs itself instead.

**Comment by blueyed** [comment]
The problem is this:
```
file /Users/Hannah/.virtualenvs/large-lots/lib/python3.5/site-packages/pytest_django/fixtures.py, line 141
  @pytest.fixture(scope='function')
  def db(request, django_db_setup, django_db_blocker):
file /Users/Hannah/hackz/datamade/large-lots/tests/lots_client/conftest.py, line 14
  @pytest.fixture
  @pytest.mark.django_db
  def django_db_setup(django_db_setup, db):
```

The `db` fixture uses `django_db_setup`, where your custom one then uses `db` already, which is not yet available.
I think pytest could be fixed to improve the error message in this case.

**Comment by blueyed** [comment]
Closing this here, will look into providing a PR to pytest to improve the error message in this case.
